extensions/jdbc/jdbc-oracle/runtime/src/main/java/io/quarkus/jdbc/oracle/runtime/RollbackOnConnectionClosePoolInterceptor.java

```diff
@@ -0,0 +1,53 @@
+package io.quarkus.jdbc.oracle.runtime;
+
+import java.sql.Connection;
+
+import org.jboss.logging.Logger;
+
+import io.agroal.api.AgroalPoolInterceptor;
+import oracle.jdbc.OracleConnection;
+
+/**
+ * Oracle has the weird behavior that it performs an implicit commit on normal connection closure (even with auto-commit
+ * disabled),
+ * which happens on application shutdown. To prevent whatever intermittent state we have during shutdown from being committed,
+ * we add an explicit rollback to each connection closure. If the connection has already received a COMMIT, the rollback will
+ * not work, which is fine.
+ * <p>
+ * The code unwraps the {@link Connection} so that we perform the rollback directly on the underlying database connection,
+ * and not on e.g. Agroal's {@link ConnectionWrapper} which can prevent the rollback from actually being executed due to some
+ * safeguards.
+ *
+ * @see https://github.com/quarkusio/quarkus/issues/36265
+ */
+public class RollbackOnConnectionClosePoolInterceptor implements AgroalPoolInterceptor {
+
+    private static final Logger LOG = Logger.getLogger(RollbackOnConnectionClosePoolInterceptor.class);
+
+    private final boolean noAutomaticRollback;
+
+    public RollbackOnConnectionClosePoolInterceptor() {
+        // if you have to use this system property, make sure you open an issue in the Quarkus tracker to explain
+        // why as we might need to adjust things
+        noAutomaticRollback = Boolean.getBoolean("quarkus-oracle-no-automatic-rollback-on-connection-close");
+    }
+
+    @Override
+    public void onConnectionDestroy(Connection connection) {
+        if (noAutomaticRollback) {
+            return;
+        }
+
+        try {
+            if (connection.unwrap(Connection.class) instanceof OracleConnection oracleConnection) {
+                if (connection.isClosed() || connection.getAutoCommit()) {
+                    return;
+                }
+
+                oracleConnection.rollback();
```

Comment:
I understand that rolling back a transaction that was already committed will just do nothing, but I wonder if it could mess with metrics... Maybe an `if (transaction in progress)` -- if possible at all -- would be safer?

Reply:
I wasn't able to figure out a way to do that. Nothing is exposed from what I can see.

There's something else to consider though: we shouldn't rollback XA connections handled by XA.

Reply:
@yrodiere I added something to avoid rolling back when we have an XA connection. I tested it with XA enabled and I enter this path properly.